lib/dataTrigger/DataTrigger.js

```diff
@@ -0,0 +1,61 @@
+const DataTriggerExecutionResult = require('./DataTriggerExecutionResult');
+const DataTriggerExecutionError = require('../errors/DataTriggerExecutionError');
+
+class DataTrigger {
+  /**
+   *
+   * @param {string} documentType
+   * @param {number} documentAction
+   * @param {function(Document, DataTriggerExecutionContext):DataTriggerExecutionResult} trigger
+   */
+  constructor(documentType, documentAction, trigger) {
+    this.documentType = documentType;
+    this.documentAction = documentAction;
+    this.trigger = trigger;
+  }
+
+  /**
+   *
+   * @param {Document} document
+   * @param {DataTriggerExecutionContext} context
+   * @returns {Promise<DataTriggerExecutionResult>}
+   */
+  async execute(document, context) {
+    const isCorrectType = document.getType() === this.documentType;
+    const isCorrectAction = document.getAction() === this.documentAction;
+    if (!isCorrectType) {
+      return new DataTriggerExecutionResult([
+        new DataTriggerExecutionError(document, context, "Document type doesn't match trigger type"),
```

Comment:
The action would be useful info in `DataTriggerExecutionError` as well.

Reply:
I would create special error types for both checks:
- `DataTriggerInvalidDocumentActionError`
- `DataTriggerInvalidDocumentTypeError`

Or we can put data trigger errors to a separate folder and remove `DataTrigger` prefix

Reply:
Well, action can't be a part of the _context_, as it's a part of the document, so you need to create a separate context for each document, which I would like to avoid. Adding action to a good idea is a good idea though :+1: